File: plugin/zeppelin/src/main/scala/jetbrains/zeppelin/utils/ThreadRun.scala
package jetbrains.zeppelin.utils

import java.util.concurrent.{Executors, TimeUnit}

import com.intellij.openapi.application.{Application, ApplicationManager}
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.util.{Computable, ThrowableComputable}

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContext, ExecutionContextExecutor, Future}
import scala.util.control.Exception.catching
import scala.util.{Failure, Success, Try}

object ThreadRun {
  private val DEFAULT_MAX_EXECUTION_TIME = 120

  def inWriteAction[T](body: => T): T = {
    val application: Application = ApplicationManager.getApplication

    if (application.isWriteAccessAllowed) {
      body
    } else {
      application.runWriteAction(
        new Computable[T] {
          def compute: T = body
        }
      )
    }
  }

  def invokeLater[T](body: => T): Unit = {
    ApplicationManager.getApplication.invokeLater(() => body)
  }

  def runWithTimeout[T](f: => T): Unit = {
    // single threaded execution context
    implicit val context: ExecutionContextExecutor = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
    Await.result(Future(f), Duration(DEFAULT_MAX_EXECUTION_TIME, TimeUnit.SECONDS))
  }

  def withProgressSynchronously[T](title: String)(body: => T): T = {
    withProgressSynchronouslyTry[T](title)(_ => body) match {
      case Success(result) => result
      case Failure(exception) => throw exception
    }
  }

  def withProgressSynchronouslyTry[T](title: String)(body: ProgressManager => T): Try[T] = {
    val manager = ProgressManager.getInstance
    catching(classOf[Exception]).withTry {
      manager.runProcessWithProgressSynchronously(new ThrowableComputable[T, Exception] {
        def compute: T = body(manager)
      }, title, false, null)
    }
  }
}

File: zeppelin/implementation/src/org/intellij/plugin/zeppelin/utils/ThreadRun.kt
package org.intellij.plugin.zeppelin.utils

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.progress.*
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
import com.intellij.openapi.project.Project

fun <T> invokeLater(body: () -> T) {
    ApplicationManager.getApplication().invokeLater { body() }
}

fun <T> withProgressSynchronously(title: String, body: () -> T): T? {
    val manager: ProgressManager = ProgressManager.getInstance()
    return try {
        return manager.runProcessWithProgressSynchronously<T, Exception>(body, title, false, null)
    } catch (e: Exception) {
        null
    }
}

fun createBgIndicator(project: Project, name: String): ProgressIndicator {
    return ProgressIndicatorProvider.getGlobalProgressIndicator() ?: BackgroundableProcessIndicator(
            project, name, PerformInBackgroundOption.ALWAYS_BACKGROUND, null, null, false)
}